packages/zoe/src/contracts/multipoolAutoswap.js

```diff
@@ -0,0 +1,627 @@
+        const liquidityExtentIn = userAllocation[liquidityKeyword].extent;
+
+        const amountMaths = zoe.getAmountMaths(liquidityKeys);
+
+        const subtract = makeSubtract(amountMaths);
+
+        const newUserAmounts = harden({
+          [CENTRAL_TOKEN]: amountMaths[CENTRAL_TOKEN].make(
+            calcExtentToRemove(
+              harden({
+                liqTokenSupply: liquidityTokenSupply,
+                poolExtent: poolAllocation[CENTRAL_TOKEN].extent,
+                liquidityExtentIn,
+              }),
+            ),
+          ),
+          [tokenKeyword]: amountMaths[tokenKeyword].make(
+            calcExtentToRemove(
+              harden({
+                liqTokenSupply: liquidityTokenSupply,
+                poolExtent: poolAllocation[tokenKeyword].extent,
+                liquidityExtentIn,
+              }),
+            ),
+          ),
+          [liquidityKeyword]: amountMaths[liquidityKeyword].getEmpty(),
+        });
+
+        const newPoolAmounts = harden({
+          [CENTRAL_TOKEN]: subtract(
+            CENTRAL_TOKEN,
+            poolAllocation,
+            newUserAmounts,
+          ),
+          [tokenKeyword]: subtract(
+            tokenKeyword,
+            poolAllocation,
+            newUserAmounts,
+          ),
+          [liquidityKeyword]: amountMaths[liquidityKeyword].add(
+            poolAllocation[liquidityKeyword],
+            amountMaths[liquidityKeyword].make(liquidityExtentIn),
+          ),
+        });
+
+        liquidityTable.update(secondaryTokenBrand, {
+          liquidityTokenSupply: liquidityTokenSupply - liquidityExtentIn,
+        });
+
+        zoe.reallocate(
+          harden([inviteHandle, poolHandle]),
+          harden([newUserAmounts, newPoolAmounts]),
+          liquidityKeys,
+        );
+        zoe.complete(harden([inviteHandle]));
+        return 'Liquidity successfully removed.';
+      },
+      swap: () => {
+        const { proposal } = zoe.getOffer(inviteHandle);
+        const getKeywordAndBrand = amountKeywordRecord => {
+          const keywords = getKeys(amountKeywordRecord);
+          if (keywords.length !== 1) {
+            rejectOffer(
+              inviteHandle,
+              `A swap requires giving one type of token for another`,
+            );
+          }
+          return harden({
+            keyword: keywords[0],
+            brand: Object.values(amountKeywordRecord)[0].brand,
+          });
+        };
+
+        const { keyword: keywordIn, brand: brandIn } = getKeywordAndBrand(
+          proposal.give,
+        );
+        const { keyword: keywordOut, brand: brandOut } = getKeywordAndBrand(
+          proposal.want,
+        );
+
+        const expected = harden({
+          give: { [keywordIn]: null },
+          want: { [keywordOut]: null },
+        });
+        rejectIfNotProposal(inviteHandle, expected);
+
+        // brandIn could either be the central token brand, or one of
+        // the secondary token brands
+
```

Comment:
And whether it ends up as two or three cases, this is certainly a situation where I would `// eslint-disable` and use
> 
> ```
>  if () {
>  } else if () {
>  } else if () {
>  }
> ```

Can you explain this more? Since the code `returns`, `else` seems unnecessary

Reply:
Even with the returns, I find 
```
 if (<condition A>) {
    .....
   return;
 } else if (<condition B>) {
    .....
   return;
 } else if (<condition C>) {
    .....
   return;
 }
```
far more readable than

```
 if (<condition A>) {
    .....
   return;
 }
 if (<condition B>) {
    .....
   return;
 }
 if (<condition C>) {
    .....
   return;
 }
```

Particularly when the bodies are long or there are more than a couple of them. If they're really exclusive, and nothing intervenes between the various bodies, then I can evaluate them separately.  If there are a series of separate conditionals, I have to look if any code has been added between conditions A and B in order to understand the context for condition C. This is also why I turn such things into switch/case when the conditions are all testing the same value--as a reader, I can make simpler assumptions about the context for each branch.

It seems like the only reason to prefer making them separate is that it makes it possible to insert code between the cases.